File: leetcode/medium/712_minimumDeleteSum.ts
function minimumDeleteSum(s1: string, s2: string): number {
    const n: number = s1.length;
    const m: number = s2.length;

    let sumOfCodes: number = 0;
    for (const char of s1) {
        sumOfCodes += char.charCodeAt(0);
    }
    for (const char of s2) {
        sumOfCodes += char.charCodeAt(0);
    }

    const substr = new Array(n).fill(0).map(() => new Array(m).fill(0));
    for (let i: number = 0; i < n; ++i) {
        for (let j: number = 0; j < m; ++j) {
            const top: number = (i > 0) ? substr[i-1][j] : 0;
            const left: number = (j > 0) ? substr[i][j-1] : 0;
            const diag: number = (i > 0 && j > 0) ? substr[i-1][j-1] : 0;

            substr[i][j] = Math.max(top, left);
            if (s1[i] === s2[j]) {
                substr[i][j] = Math.max(substr[i][j], diag + s1[i].charCodeAt(0));
            }
        }
    }

    return sumOfCodes - (substr[n-1][m-1] * 2);
};

/*

  3 3 4 1 1 2 8 9
1 0 0 0 1 1 1 1 1
1 0 0 0 1 2 2 2 2
2 0 0 0 1 2 3 3 3
6 0 0 0 1 2 3 3 3
3 1 1 1 1 2 3 3 3
5 1 1 1 1 2 3 3 3
3 1 2 2 2 2 3 3 3
4 1 2 3 3 3 3 3 3


*/

const test = () => {
    const params = [
        {
            input: ["sea", "eat"],
            output: 231,
        },
        {
            input: ["delete", "leet"],
            output: 403,
        },
    ];

    params.forEach(({input, output}) => {
        const [s1, s2]: string[] = input;
        const result = minimumDeleteSum(s1, s2);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();

File: leetcode/medium/712_minimumDeleteSum.py
from typing import List
import heapq
import math
from collections import defaultdict, deque

class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        N = len(s1)
        M = len(s2)

        total = 0
        for i in range(N):
            total += ord(s1[i])
        for i in range(M):
            total += ord(s2[i])

        memo = [[-1]*M for i in range(N)]

        def dfs(id1, id2):
            if id1 == N or id2 == M:
                return 0

            if memo[id1][id2] != -1:
                return memo[id1][id2]

            res = 0
            if s1[id1] == s2[id2]:
                res += ord(s1[id1])
                res += dfs(id1+1, id2+1)

            r1 = dfs(id1+1, id2)
            if r1 > res:
                res = r1
            r1 = dfs(id1, id2+1)
            if r1 > res:
                res = r1

            memo[id1][id2] = res

            return res

        subSrtSum = dfs(0, 0)

        return total - subSrtSum*2

'''
- s e a
e 0 1 1
a 0 1 2
t 0 1 2
'''

def test ():
    params = [
        {
            'input': ["sea", "eat"],
            'output': 231,
        },
        {
            'input': ["delete", "leet"],
            'output': 403,
        },
    ]
    solution = Solution()

    for param in params:
        s1, s2 = param['input']
        result = solution.minimumDeleteSum(s1, s2)
        print(
            'SUCCESS' if result == param['output'] else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )


if __name__ == '__main__':
    test()
